pkg/controllers/management/secretmigrator/authconfig.go

```diff
@@ -0,0 +1,158 @@
+package secretmigrator
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/rancher/norman/condition"
+	"github.com/rancher/norman/objectclient"
+	apimgmtv3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/auth/providers/common"
+	client "github.com/rancher/rancher/pkg/client/generated/management/v3"
+	"github.com/rancher/rancher/pkg/namespace"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	serviceAccountPasswordFieldName = "serviceaccountpassword"
+	authConfigKind                  = "authconfig"
+)
+
+// syncAuthConfig syncs the authentication config and removes/migrates secrets as needed.
+func (h *handler) syncAuthConfig(_ string, authConfig *apimgmtv3.AuthConfig) (runtime.Object, error) {
+	if authConfig == nil || !authConfig.Enabled || apimgmtv3.AuthConfigConditionSecretsMigrated.IsTrue(authConfig) {
+		return authConfig, nil
+	}
+
+	if authConfig.Type != client.ShibbolethConfigType {
+		unstructuredConfig, err := getUnstructuredAuthConfig(h.authConfigs, authConfig)
+		if err != nil {
+			return nil, err
+		}
+		newUnstructuredConfig, err := setUnstructuredStatus(unstructuredConfig, apimgmtv3.AuthConfigConditionSecretsMigrated, "True")
+		if err != nil {
+			return nil, fmt.Errorf("failed to set the status on unstructured AuthConfig %s: %w", authConfig.Name, err)
+		}
+
+		updated, err := h.authConfigs.Update(authConfig.Name, newUnstructuredConfig)
+		if err != nil {
+			return nil, fmt.Errorf("unable to update migration status of authconfig: %w", err)
+		}
+		return updated, nil
+	}
+
+	updated, err := apimgmtv3.AuthConfigConditionSecretsMigrated.DoUntilTrue(authConfig, func() (runtime.Object, error) {
```

Comment:
I'm a bit nervous about this code, and have a few questions on it:
- First, does this update the secretsMigrated condition when the function returns nil?
- Second, does it update on the unstructured object, or on the structured object? If on the structured object, does this run the risk of removing any active configuration?

Reply:
It looks like the migrated condition will not be set on a nil return, but it will be retried.
It is doing its work on the unstructured object and as far as I can tell, no fields are dropped.

Reply:
To make sure that I understand this correct - it looks like "migrateShibbolethSecrets" decodes to (and returns) a `ShibbolethConfig`. Since this is a structured value, our Norman stuff can update the condition on this value without dropping any fields. We can't use this strategy for the other types, since we don't know what the exact subtype of that authconfig is (and it would be significantly more code to figure it out). 

Therefore, though this works, we can't adopt other areas to use this and can only do it for this part (where we know what type the config is). Do I have that right?

If so, I would recommend changing the return type of the "migrateShibbolethSecrets" function to `*apimgmtv3.ShibbolethConfig` - this should keep your code relatively the same and will give us better protection against bad changes to this function in the future (maybe also include a comment on why it is important that this function returns a well-structured config too).

Reply:
Yes, I think you've nailed it.
I like the changing of the return type of that function.  Making it specific makes it unlikely that it'll be used (or attempted) by any other similar operation in the future.  I'll make that change when I rebase.